tasks.upgrader.js

```diff
@@ -18,25 +20,36 @@ const structureTypesToName = {
 }
 
 const upgrader = {
-    createCS: function (sources, structureType, room, number) {
-
-        const x = sources.pos ? sources.pos.x : false
-        const y = sources.pos ? sources.pos.y : false
+    createCS: function (structureType, room, numberOfStructures) {
 
-        if (!x || !y) {
-            console.log('CreateCSError: Invalid x,y coordinates', JSON.stringify({x,y}))
-            return
-        }
         const name = room.name + structureTypesToName[structureType]
-        if (number === 1 || number === null) {
-            const structures = room.find(FIND_STRUCTURES)
-            if (structureType === STRUCTURE_CONTAINERS) {
-                const containers = structures.filter(structure => structure.structureType === STRUCTURE_CONTAINERS)
+        if ((numberOfStructures === 1 || numberOfStructures === null) && (structureType == STRUCTURE_CONTAINER)) {
+
+            // Merged array of SPAWNS and SOURCES
+            const structures = room.find(FIND_SOURCES).concat(room.find(FIND_SPAWNS))
+
+            // Create container near SPAWN or SOURCE
+            for (const structure in structures) {
+                let pos = utils.construction.findStructurePosition(structure) // pos = [x, y]
+                if (pos === null) {
+                    console.log('createCSError: failed to create source', JSON.stringify({ source: structure }))
+                    return
+                } else {
+                    let [x, y] = pos
+                    const validPosition = utils.construction.findValidPositionNearby(x, y)
+                    if (validPosition === null) {
+                        console.log(`createCSError: utils.construction.findValidPositionNearby returned ${validPosition} `, JSON.stringify({ source: structure }))
+                        return
+                    } else {
+                        const [_x, _y] = validPosition
+                        Game.rooms.sim.createConstructionSite(_x, _y, structureType,
+                            name + number.toString());
```

Comment:
Is this meant to be `numberOfStructures`?
```suggestion
                            name + numberOfStructures.toString());
```

Reply:
Good catch. Thanks!